factor/email/classes/factor.php

```diff
@@ -31,4 +32,71 @@
 
 class factor extends object_factor_base {
 
+    public function login_form_definition($mform) {
+        $userfactors = $this->get_enabled_user_factors();
+
+        if (count($userfactors) > 0) {
+            $mform->addElement('hidden', 'secret');
+            $mform->setType('secret', PARAM_ALPHANUM);
+
+            $mform->addElement('text', 'verificationcode', get_string('verificationcode', 'factor_email'));
+            $mform->addRule('verificationcode', get_string('required'), 'required', null, 'client');
+            $mform->setType("verificationcode", PARAM_ALPHANUM);
+        }
+
+        return $mform;
+    }
+
+    public function login_form_definition_after_data($mform) {
+        $secretfield = $mform->getElement('secret');
+        $secret = $secretfield->getValue();
+
+        if (empty($secret)) {
+            $secret = random_int(100000, 999999);
+            $secretfield->setValue($secret);
+            $this->email_secret_code($secret);
+        }
+        return $mform;
+    }
+
+    public function email_secret_code($secret) {
+        global $USER;
+        $noreplyuser = \core_user::get_noreply_user();
+        $subject = get_string('email:subject', 'factor_email');
+        $message = get_string('email:message', 'factor_email', $secret);
+        $messagehtml = text_to_html($message);
+        email_to_user($USER, $noreplyuser, $subject, $message, $messagehtml);
+    }
+
+    public function login_form_validation($data) {
+        $return = array();
+
+        if ($data['verificationcode'] != $data['secret']) {
+            $return['verificationcode'] =  'Wrong verification code';
+        }
+
+        return $return;
+    }
+
+    public function get_all_user_factors() {
+        global $USER;
+
+        $id = 1;
+        $name = $this->name;
+        $useremail = $USER->email;
+        $timemodified = '';
+        $timecreated = '';
+        $disabled = (int)!$this->is_enabled();
+
+        $return = array();
+        $return[1] = new \stdClass();
+        $return[1]->id = $id;
+        $return[1]->name = $name;
+        $return[1]->useremail = $useremail;
+        $return[1]->timemodified = $timemodified;
+        $return[1]->timecreated = $timecreated;
+        $return[1]->disabled = $disabled;
```

Comment:
This seems a bit redundant? why not just

```
$return = array(
  'id' => 1,
  'name' => $this->name
...
);
```

Also I would mix in $USER->confirmed as part of the enabled

Reply:
We need other fields to build a user preference table. Also we user `disabled` property in `get_enabled_user_factors()` method of factor base class

Reply:
the first comment was purely a code formatting issue, it is 14 lines of code which could be 6 lines of code and do the same thing and be easier to read.

Reply:
It needs to be an array of objects, not an array of arrays.
However, it can be simplified. Thanks

Reply:
You can cast arrays to objects like this:

```
$return = array((object) [
  'id' => 1,
  'name' => $this->name,
  'useremail' => $useremail,
   ...
]);
```

https://stackoverflow.com/questions/19272011/how-to-convert-an-array-into-an-object-using-stdclass/19272047